File: cases/taint/scala/anonymous-implementation-positive/AnonymousImplementationPositive.scala
package dataflowbench

object AnonymousImplementationPositive {
  trait Handler {
    def handle(value: String): Unit
  }

  def dfb_source(): String = { // DFB-SOURCE: anonymous-implementation-input
    "tainted"
  }

  def dfb_sink(value: String): Unit = {} // DFB-SINK: anonymous-implementation-sink

  def run(): Unit = {
    val leak: Handler = new Handler { // DFB-WITNESS: anonymous-implementation-bind
      def handle(value: String): Unit = {
        dfb_sink(value)
      }
    }
    val drop: Handler = new Handler {
      def handle(value: String): Unit = {
        dfb_sink("clean")
      }
    }
    leak.handle(dfb_source())
  }
}

File: cases/taint/kotlin/anonymous-implementation-positive/AnonymousImplementationPositive.kt
package dataflowbench

object AnonymousImplementationPositive {
    fun interface Handler {
        fun handle(value: String)
    }

    fun dfb_source(): String { // DFB-SOURCE: anonymous-implementation-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: anonymous-implementation-sink

    fun run() {
        val leak: Handler = object : Handler { // DFB-WITNESS: anonymous-implementation-bind
            override fun handle(value: String) {
                dfb_sink(value)
            }
        }
        val drop: Handler = object : Handler {
            override fun handle(value: String) {
                dfb_sink("clean")
            }
        }
        leak.handle(dfb_source())
    }
}
